### PR TITLE
[Features] hang up on call

### PR DESCRIPTION
**Pull request by ahammer**
## Purpose

* Hang up on call accepted

## Does this introduce a breaking change?

```
[] Yes
[X] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[X] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test

* Accept the "Read phone state" permission (should show on setup screen)
* Join a Teams/Groups meeting
* Receive & Accept a PTSN (I.e. regular phone call)
* Observe hangup of Teams/Groups meeting

## What to Check
Verify that the following are valid
- When phone_state granted, calls hang up on incoming accepted call
- When phone_state denied, hang up feature gracefully degrades (app is stable, but no hang up).

### CODE REVIEW

**Comment by phileo99** [comment]
Can we add some unit tests?

**Comment by ahammer** [comment]
> Can we add some unit tests?

I changed the structure a bit, but I can add a few tests for the new redux state and actions, will bring them in shortly.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    fun onOffHook() {
+        if (store.getCurrentState().callState.callingStatus == CallingStatus.CONNECTED) {
+            store.dispatch(CallingAction.CallEndRequested())
+        }
+    }
+}
+
+// / Legacy Version for < 31
+private class DetectPhoneCallLegacy(private val telephonyManager: TelephonyManager, override val store: Store<ReduxState>) :
+    PhoneStateListener(), DetectPhoneCall {
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt:62]: `PhoneStateListener` is deprecated, can we add `@Suppress("DEPRECATION")`

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+    override fun onCallStateChanged(state: Int, phoneNumber: String?) {
+        super.onCallStateChanged(state, phoneNumber)
+        if (state == TelephonyManager.CALL_STATE_OFFHOOK) {
+            onOffHook()
+        }
+    }
+
+    override fun register() {
+        telephonyManager.listen(this, PhoneStateListener.LISTEN_CALL_STATE)
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt:72]: `PhoneStateListener` package already `import`'d, so this qualifier name is not needed

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt [hunk 3]
```diff
@@ -111,34 +148,10 @@ internal class PermissionManager(
...
     }
 
+    fun setPhonePermissionState() {
+        store.dispatch(PermissionAction.PhonePermissionIsSet(getPhonePermissionState(activity)))
+    }
+
     private fun isPermissionGranted(permission: String) =
         ActivityCompat.checkSelfPermission(activity, permission) == PackageManager.PERMISSION_GRANTED
-
-    private fun getPermissionsList(): Array<String> =
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:117]: Just wondering why was this removed?

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt [hunk 4]
```diff
@@ -9,6 +9,8 @@ internal sealed class PermissionAction :
     Action {
     class AudioPermissionRequested : PermissionAction()
     class CameraPermissionRequested : PermissionAction()
+    class PhonePermissionRequested : PermissionAction()
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt:12]: These are all stateless, can we convert them to sealed objects instead

**Review by phileo99 — CHANGES_REQUESTED** [review]
If I deny phone permission, then later on allow, then  the hangup on call doesn't work

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt [hunk 5]
```diff
@@ -58,16 +70,26 @@ internal class PermissionManager(
         }
     }
 
+    private fun createPhonePermissionRequest() {
+        if (getPhonePermissionState(activity) == PermissionStatus.NOT_ASKED) {
+            phonePermissionLauncher.launch(Manifest.permission.READ_PHONE_STATE)
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:75]: Doesn't this need to be added to the Android manifest ?

Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:75]: Doesn't this need to be added to the Android manifest ?

Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:75]: It is in there as well, but it's a runtime permission on newer version of android, compile time old versions.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt (hunk 4 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt]: Converted

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt (hunk 3 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:117]: IDE said it was unused so I deleted.

We have a list of permissions for audio, and a single permission for camera/phone_state permission requests. I think this is just old dead code.

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt (hunk 2 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt]: done

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt (hunk 1 above)
Comment (ahammer) [reply azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/DetectPhoneCallManager.kt:63]: Done

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt [hunk 6]
```diff
@@ -7,8 +7,10 @@ import com.azure.android.communication.ui.calling.redux.state.PermissionStatus
 
 internal sealed class PermissionAction :
     Action {
-    class AudioPermissionRequested : PermissionAction()
-    class CameraPermissionRequested : PermissionAction()
-    class AudioPermissionIsSet(val permissionState: PermissionStatus) : PermissionAction()
-    class CameraPermissionIsSet(val permissionState: PermissionStatus) : PermissionAction()
+    sealed class AudioPermissionRequested : PermissionAction()
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt:10]: replace `sealed class` with `object` for the stateless classes

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt [hunk 7]
```diff
@@ -7,8 +7,10 @@ import com.azure.android.communication.ui.calling.redux.state.PermissionStatus
...
+    sealed class CameraPermissionRequested : PermissionAction()
+    sealed class PhonePermissionRequested : PermissionAction()
+    sealed class AudioPermissionIsSet(val permissionState: PermissionStatus) : PermissionAction()
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/redux/action/PermissionAction.kt:13]: remove `sealed` for the stateful classes

File: azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt [hunk 8]
```diff
@@ -22,18 +23,29 @@ internal class PermissionManager(
...
 
-    private val audioPermission by lazy { getPermissionsList() }
+    private val audioPermissions =
+        arrayOf(
+            Manifest.permission.BLUETOOTH_CONNECT,
+            Manifest.permission.RECORD_AUDIO,
+            Manifest.permission.ACCESS_NETWORK_STATE,
+            Manifest.permission.WAKE_LOCK,
+            Manifest.permission.MODIFY_AUDIO_SETTINGS,
+            Manifest.permission.FOREGROUND_SERVICE
```
Comment (phileo99) [inline azure-communication-ui/azure-communication-ui/src/calling/java/com/azure/android/communication/ui/calling/presentation/manager/PermissionManager.kt:36]: `FOREGROUND_SERVICE` permission available only in API28 and newer
`BLUETOOTH_CONNECT` permissioin available only in API31 and newer

**Review by phileo99 — CHANGES_REQUESTED** [review]
denying phone state permissions is still causing issues.
Did you want to deal with that in this PR or in a separate PR ?
